### DISCUSSION TITLE
Why items are moved without user input

### DISCUSSION THREAD

**Discussion by jordanbaird** [jordanbaird/Ice#298] [General]
There have been a lot of bug reports about Ice moving menu bar items without the user explicitly moving them. This includes items being moved from one section to another, most commonly the always-hidden section. I'm writing this post to explain why this isn't a bug with Ice, why it happens, and what Ice will eventually be able to do about it.

Most apps remember the positions of their menu bar items, but some don't, causing their menu bar items to be treated as new items every time the app is launched. macOS adds new menu bar items to the far left of the menu bar, which is also where the always-hidden section is located. In rare cases, there might be multiple items with conflicting remembered positions. If this happens, it is again up to macOS to decide how to handle it.

In short, Ice doesn't have anything to do with why these items are being moved. It comes down to a macOS behavior. Once #26 is implemented, Ice will be able to remember the position of every item and ensure that they remain in the correct order.

**Comment by jdwhite** [comment]
Thank you for the informative explanation.

**Comment by joostiphone** [comment]
Thanks for the explanation @jordanbaird . Is there any feature scheduled on short term to "fix" this? Bartender has a solution for this and remembers the position. I prefer Ice above Bartender because of Ice's being lean, but this issue holds me from stepping away from Bartender.

**Reply by jordanbaird** [reply]
The only real solution is going to be to implement some sort of auto-arranging based on a stored configuration, which is basically what a profile is, so there is no intermediate fix planned.

**Comment by tadg-paul** [comment]
Ah right, I was kind of hoping there might be some way to let new icons default to visible, but it sounds like a limitation of how macOS is handling them.

Which makes me think Profiles if you can figure it out is some sort of black magic!

My problem is with Hammerspoon creating menu icons as brand new on every launch ... if anyone has any ideas on how to fix that ... no maybe I shouild go raise an issue on their github actually and stop pestering you here haha

**Reply by jordanbaird** [reply]
Not quite black magic haha. It's just a matter of keeping track of the order of items and moving them back into position if needed. The main reason I haven't implemented it yet is just because it's a pretty big task, but I'm starting on it now as the next major feature.

**Reply by tino** [reply]
Just out of curiosity, what are you currently storing/tracking to maintain the preferences of what goes where? (I'm a programmer, but not versatile in Swift and couldn't find it with a quick browse around)

**Reply by Rhys-T** [reply]
For Hammerspoon in particular, you might try using the `:autosaveName(...)` method on your `hs.menubar` instances to assign each one a different name, so that Hammerspoon can recognize it as the same menu after a restart and restore its position properly.

**Reply by FezVrasta** [reply]
Has there been any progress on this?
